es6/util.js

```diff
@@ -317,3 +396,140 @@ export function applyPatches(obj, patches, flatten_styles=true) {
+        }
+    }
+
+    return anim_queue.filter((anim) => is_between(anim))
+}
+
+
+// these are no longer useful -- they only make sense with unidirectional time
+export const pastAnimations = (anim_queue, timestamp) =>
+    anim_queue.filter(({start_time, duration}) =>
+        (start_time + duration < timestamp))
+
+export const futureAnimations = (anim_queue, timestamp) =>
+    anim_queue.filter(({start_time, duration}) => (start_time > timestamp))
+
+// export const sortedAnimations = (anim_queue) => {
+//     return [...anim_queue].sort((a, b) => {
+//         // sort by end time, if both are the same, sort by start time,
+//         //  and properly handle infinity
+//         if (a.end_time == b.end_time) {
+//             return b.start_time - a.start_time
+//         } else {
+//             if (a.end_time == Infinity) {
+//                 return 1
+//             }
+//             else if (b.end_time == Infinity) {
+//                 return -1
+//             }
+//             else {
+//                 return b.end_time - a.end_time
+//             }
+//         }
+//     })
+// }
+
+// 0 /a /b /c       3
+// 1 /a /b          2
+// 2 /a /b /e /d    4
+
+const parentExists = (paths, path) => {
+    let parent = ''
+    for (let key of path.split('/').slice(1)) {   // O(path.length)
+        parent = `${parent}/${key}`
+        if (paths.has(parent)) {
+            return true
+        }
+    }
+    return false
+}
+
+export const uniqueAnimations = (anim_queue) => {
+    const paths = new Set()
+    const uniq_anims = []
+
+    for (let anim of reversed(anim_queue)) {    // O(anim_que.length)
+        if (!parentExists(paths, anim.path)) {
+            uniq_anims.push(anim)
+            paths.add(anim.path)
+        }
+    }
+
+    return uniq_anims.reverse()
+}
+
+export const activeAnimations = ({anim_queue, warped_time,
+                                  former_time, uniqueify}) => {
+    if (warped_time === undefined || former_time === undefined) {
+        throw 'Both warped_time and former_time must be passed to get activeAnimations'
+    }
+
+    let anims = [
+        ...finalFrameAnimations({anim_queue, former_time, warped_time}),
+        ...currentAnimations({anim_queue, warped_time}),
+    ]
+
+    if (uniqueify)
+        anims = uniqueAnimations(anims)
+
+    return anims
+}
+
+export const computeAnimatedState = (anim_queue, warped_time,
+        former_time=null) => {
+    former_time = former_time === null ? warped_time : former_time
+
+    const active_animations = activeAnimations({anim_queue,
+                                                warped_time,
+                                                former_time,
+                                                uniqueify: false})
+    let patches = []
+    // console.log({active_animations})
+    for (let animation of active_animations) {
+        try {
+            const delta = warped_time - animation.start_time
+            if (animation.merge) {
```

Comment:
Can this whole if/else block be inside a function called `getPatches(animation, ...)`?